contents/c++/rvalue.md

```diff
@@ -0,0 +1,145 @@
+
+However, [1 major flaw](https://www.quora.com/What-are-the-drawbacks-of-pass-by-value-result) of Value Semantics and passing-by-value is having poor performance and scability. When we wish to only read the values without modifying it, creating an additional copy just for this purpose is an unnecessary computation, especially if we scale up to larger objects.
+
+### Reference Semantics
+Reference (or pointer) semantics is another mode available for users in C++. C++ allows users to declare [pointers](http://www.cplusplus.com/doc/tutorial/pointers/) and [references](https://www.geeksforgeeks.org/references-in-c/) to the address of the value. As such, we can pass around these pointers to functions and objects, and all of them will refer to and use the same value (address).
+
+One large advantage of using Reference Semantics is that it allows us pass around large objects without having to create an additional copy, thus improving on the performance and scability of the program. It also allows objects to interact with each other as they can have pointers to and modify the same value.
+
+However, Reference Semantics also comes with its own problems. In a multi-threaded environment, having such object pointing to the same value will lead to unnecessary data races and un-deterministic behaviour, and requires extra work in trying to synchronize the values between objects. More details of this can be found in the [Java Concurrency](java/JavaConcurrency.md) and [Java Synchronization](java/JavaSynchronization.md).
+Reference Semantics and also lead to unintended behaviour if we modify the values in 1 object, which in turn affects the other objects which have pointers to the same values as well, an example shown in this [article](http://www.drdobbs.com/cpp/optimization-calling-by-value-or-by-refe/232400151).
+
+## Move Semantics
+As mentioned above, 1 major flaw of Value Semantics is that we will create a copy of the value, which can computationally expensive if this value is a large object. However, we may also not want to use Reference Semantics due to its different problems as discussed above. As such, in order to continue using Value Semantics without creating this unnecessary copy, C++ introduced a new mode, Move Semantics.
+
+### Rvalue and Lvalue references
+To understand how Move Semantics work in C++, we first need to know about the distinction between rvalue and lvalue references. 1 simple line to describe the differencce between these 2 is:
+```
+lvalue = rvalue
+```
+where lvalue (left value) is on the left side of the equation, and rvalue (right value) is on the right.
```

Comment:
**From the example above, the** lvalue (left value) is on the left side of the equation...

Not sure if this sentence is required though. It just states what the code looks like.